### PR TITLE
Create an LSP testing and assertion framework

### LINKED ISSUE #326: Create an LSP testing and assertion framework
Currently, testing the LSP is tricky to test which presents an obvious gap in testing and quality. Therefore, we need to create/identify some functions, patterns, or existing tools that will vastly improve the ability to test the LSP features.

## AC:
- [x] The testing framework is capable of easily creating and managing virtual/temporary documents
- [x] The testing framework provides a clear, simplified API for triggering events as the LSP client
- [x] The testing framework provides common assertion functions
  - [x] Abstract away the need to dig through LSP responses
- [x] Create a unit test class that handles the setup/tear down associated with LSP tests

### PR DESCRIPTION
**Pull request by cvchaparro**
Closes #326

Changes:
- Add some infrastructure to better enable LSP functionality testing.
- Add a base test case for LSP tests that adds functionality for handling text documents, sending specific requests, etc.
- Add a general base test case that handles common set up (e.g. reloading the active context, etc).
- Add some basic LSP tests for hover and completion.

### CODE REVIEW

File: python/tests/lang/test_server.py [hunk 1]
```diff
@@ -1,45 +1,126 @@
...
+
+
+class TestLspServer(BaseLspTestCase, IsolatedAsyncioTestCase):
+    async def asyncSetUp(self) -> None:
+        await super().asyncSetUp()
+        self.active_context = self.client.server.language_context
+        await self.create_document("test.aac", TEST_DOCUMENT_CONTENT)
+
+    async def test_adds_definitions_when_opening_file(self):
+        await self.create_document("added.aac", TEST_DOCUMENT_ADDITIONAL_CONTENT)
```
Comment (Coffee2Bits) [inline python/tests/lang/test_server.py:17]: I like the pattern of demonstrating that the definitions aren't in the context prior to being piped in via the LSP.
```suggestion
        self.assertIsNone(self.active_context.get_definition_by_name(TEST_ADDITIONAL_SCHEMA_NAME))             
        self.assertIsNone(self.active_context.get_definition_by_name(TEST_ADDITIONAL_MODEL_NAME))
        
        await self.create_document("added.aac", TEST_DOCUMENT_ADDITIONAL_CONTENT)
```

**Review by Coffee2Bits — CHANGES_REQUESTED** [review]
Looks good; I'm excited by how easy those LSP tests look now! 

I did leave some feedback and suggestions.

File: python/tests/lang/base_lsp_test_case.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    document: Optional[TextDocument] = None
+
+    async def asyncSetUp(self) -> None:
+        await super().asyncSetUp()
+
+        self.client = LspTestClient()
+        await self.client.start()
+        await self.client.send_request(
+            methods.INITIALIZE,
+            InitializeParams(process_id=12345, root_uri=self.to_uri("root.aac"), capabilities=ClientCapabilities()),
```
[the remark this answers is not in the document]
Comment (cvchaparro) [reply python/tests/lang/base_lsp_test_case.py:32]: It's interesting that this sparked me to look at the [`InitializeParams`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeParams) structure in the LSP spec and I see now that `root_uri` is deprecated in favor of `workspaceFolders`. I will replace it with that so we don't forget but I'm curious, now, where this even gets used? I seem to recall it being required at one point but we never did anything with it, to my knowledge.

**Review by Coffee2Bits — APPROVED** [review]
Good work!

**Pull request merged**